pg_replicate/src/clients/delta.rs

```diff
@@ -0,0 +1,235 @@
+impl DeltaClient {
+    fn naive_date_to_arrow(date: NaiveDate) -> i32 {
+        let epoch = NaiveDate::from_ymd_opt(1970, 1, 1).unwrap();
+        (date - epoch).num_days() as i32
+    }
+
+    fn postgres_to_delta(typ: &Type) -> DataType {
+        match typ {
+            &Type::BOOL => DataType::BOOLEAN,
+            &Type::CHAR | &Type::BPCHAR | &Type::VARCHAR | &Type::NAME | &Type::TEXT => {
+                DataType::STRING
+            }
+            &Type::INT2 | &Type::INT4 | &Type::INT8 => DataType::INTEGER,
+            &Type::FLOAT4 | &Type::FLOAT8 | &Type::NUMERIC => DataType::FLOAT,
+            &Type::DATE => DataType::DATE,
+            &Type::TIME | &Type::TIMESTAMP | &Type::TIMESTAMPTZ => DataType::TIMESTAMP,
+            &Type::UUID => DataType::STRING,
+            &Type::OID => DataType::INTEGER,
+            &Type::BYTEA => DataType::BYTE,
+            _ => DataType::STRING,
+        }
+    }
+
+    fn postgres_to_arrow(typ: &Type) -> ArrowDataType {
+        match typ {
+            &Type::BOOL => ArrowDataType::Boolean,
+            &Type::CHAR | &Type::BPCHAR | &Type::VARCHAR | &Type::NAME | &Type::TEXT => {
+                ArrowDataType::Utf8
+            }
+            &Type::INT2 | &Type::INT4 | &Type::INT8 => ArrowDataType::Int32,
+            &Type::FLOAT4 | &Type::FLOAT8 | &Type::NUMERIC => ArrowDataType::Float16,
+            &Type::DATE => ArrowDataType::Date32,
+            &Type::TIME | &Type::TIMESTAMP | &Type::TIMESTAMPTZ => ArrowDataType::Utf8,
+            &Type::UUID => ArrowDataType::Utf8,
+            &Type::OID => ArrowDataType::Utf8,
+            _ => ArrowDataType::Utf8,
+        }
+    }
+
+    pub async fn create_table_schema(
+        &mut self,
+        tb_name: String,
+        columns: &[ColumnSchema],
+    ) -> Result<Arc<Schema>, DeltaTableError> {
+        let full_path = self.delta_full_path(tb_name.clone());
+
+        if Path::new(&full_path).join("_delta_log").exists() {
+            DeltaOps::try_from_uri(&full_path).await?.delete().await?;
+        }
+
+        let mut table = DeltaOps::try_from_uri(&full_path)
+            .await?
+            .create()
+            .with_table_name(tb_name.clone());
+
+        let arrow_schema = Self::generate_schema(columns, &mut table)?;
+
+        Ok(arrow_schema)
+    }
+
+    fn generate_schema(
+        columns: &[ColumnSchema],
+        table: &mut CreateBuilder,
+    ) -> Result<Arc<Schema>, DeltaTableError> {
+        let mut schema: Vec<Field> = vec![];
+
+        for column in columns {
+            *table = table.clone().with_column(
+                column.name.as_str(),
+                Self::postgres_to_delta(&column.typ),
+                false,
+                None,
+            );
+
+            schema.push(Field::new(
+                column.name.as_str(),
+                Self::postgres_to_arrow(&column.typ),
+                true,
+            ));
+        }
+
+        schema.push(Field::new("OP", ArrowDataType::Utf8, true));
```

Comment:
What does the "OP" field do?

Reply:
this one for capture the ("INSERT","DELETE","UPDATE") process

Reply:
![Screenshot 2024-12-02 at 7 49 19 PM](https://github.com/user-attachments/assets/017a01c2-f9b8-411b-8a08-32132ab636a4)

will be like this